File: Advanced/PageAndRDL/CustomTileProvider/VB.NET/CustomTileProviders/WebRequestHelper.vb
Imports System.IO
Imports System.Net
Imports System.Runtime.CompilerServices

Friend Module WebRequestHelper
    ''' <summary>
    ''' Load raw data into MemoryStream from specified Url.
    ''' </summary>
    ''' <paramname="url">Data source Url</param>
    ''' <paramname="timeoutMilliseconds">Timeout in milliseconds. If -1 the default timeout will  used.</param>
    ''' <paramname="success">Success callback.</param>
    ''' <paramname="error">Error callback.</param>
    ''' <paramname="userAgent">User-Agent for request.</param>
    ''' <returns></returns>
    Public Sub DownloadDataAsync(ByVal url As String, ByVal timeoutMilliseconds As Integer, ByVal success As Action(Of MemoryStream, String), ByVal [error] As Action(Of Exception), ByVal Optional userAgent As String = Nothing)
    Dim request = WebRequest.CreateHttp(url)

    If Not String.IsNullOrEmpty(userAgent) Then request.UserAgent = userAgent

    If timeoutMilliseconds > 0 Then
        request.Timeout = timeoutMilliseconds
    End If

    request.BeginGetResponse(Sub(ar)
                                 Try
                                     Dim response = request.EndGetResponse(ar)

                                     'Copy data from buffer (It must be done, otherwise the buffer overflow and we stop to receive responses).
                                     Dim stream = New MemoryStream()
                                     Dim responseStream = response.GetResponseStream()
                                     If responseStream IsNot Nothing Then
                                         responseStream.CopyTo(stream)
                                         success(stream, response.ContentType)
                                     Else
                                         [error](New NullReferenceException(NameOf(responseStream)))
                                     End If
                                 Catch exception As Exception
                                     [error](exception)
                                 End Try
                             End Sub, Nothing)

End Sub
End Module

File: Advanced/PageAndRDL/CustomTileProvider/C#/CustomTileProviders/WebRequestHelper.cs
using System;
using System.IO;
using System.Net;
using System.Threading.Tasks;

namespace GrapeCity.ActiveReports.Samples.CustomTileProviders
{
	internal static class WebRequestHelper
	{
		/// <summary>
		/// Load raw data into MemoryStream from specified Url.
		/// </summary>
		/// <param name="url">Data source Url</param>
		/// <param name="timeoutMilliseconds">Timeout in milliseconds. If -1 the default timeout will  used.</param>
		/// <param name="success">Success callback.</param>
		/// <param name="error">Error callback.</param>
		/// <param name="userAgent">User-Agent for request.</param>
		/// <returns></returns>
		public static void DownloadDataAsync(string url, int timeoutMilliseconds, Action<MemoryStream, string> success, Action<Exception> error, string userAgent = null)
		{
			var request = WebRequest.CreateHttp(url);

			if (!string.IsNullOrEmpty(userAgent))
				request.UserAgent = userAgent;

			if (timeoutMilliseconds > 0)
			{
				request.Timeout = timeoutMilliseconds;
			}

			request.BeginGetResponse(ar =>
			{
				try
				{
					var response = request.EndGetResponse(ar);

					//Copy data from buffer (It must be done, otherwise the buffer overflow and we stop to receive responses).
					var stream = new MemoryStream();
					var responseStream = response.GetResponseStream();
					if (responseStream != null)
					{
						responseStream.CopyTo(stream);
						success(stream, response.ContentType);
					}
					else
					{
						error(new NullReferenceException(nameof(responseStream)));
					}
				}
				catch (Exception exception)
				{
					error(exception);
				}
			}, null);

		}
	}
}
